### PR TITLE
feat: migration guide

### PR DESCRIPTION
**Pull request by uF4No**
# Description

More detailed migration guide for Hardhat projects

## Linked Issues

N/A

### CODE REVIEW

File: content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md [hunk 1]
```diff
@@ -1,275 +1,527 @@
...
-## Compile contracts
+```typescript [hardhat-zksync-ethers]
+const greeter = await hre.ethers.deployContract('Greeter', ['Hi there!']);
+await greeter.waitForDeployment();
+```
 
-To compile your contracts for ZKsync Era, run:
+::
+
+In-case a custom deployment is needed, use `ContractFactory`.
```
Comment (itsacoyote) [inline content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md:287]:
```suggestion
When a custom deployment is needed, use `ContractFactory`.
```

File: content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md [hunk 2]
```diff
@@ -1,275 +1,527 @@
...
+const greeterContract = GreeterFactory.deploy(); // if any, pass constructor arguments in deploy arguments
+await greeter.waitForDeployment();
 ```
 
-```bash [npm]
-npm hardhat compile  --network zkSyncTestnet
+```typescript [hardhat-zksync-ethers]
+const GreeterFactory = await hre.ethers.getContractFactory('Greeter');
+const greeterContract = GreeterFactory.deploy(); // if any, pass constructor arguments in deploy arguments
+await greeter.waitForDeployment();
```
Comment (itsacoyote) [inline content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md:300]: All of these code examples have the same code between the hardhat-ethers and hardhat-zksync-ethers. Is that expected?

File: content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md [hunk 3]
```diff
@@ -1,275 +1,527 @@
...
+```
+
+```typescript [hardhat-zksync-upgradeable]
+const constructorArguments = [...];
+const initializerFunctionName = 'initialize'
+const boxFactory = await hre.ethers.getContractFactory("Box");
+const box = await hre.upgrades.deployProxy(boxFactory, constructorArguments, {
+  initializer: initializerFunctionName
+});
+await box.waitForDeployment();
```
Comment (itsacoyote) [inline content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md:356]: Are these supposed to be the exact same?

File: content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md [hunk 4]
```diff
@@ -1,275 +1,527 @@
...
+const boxV2 = await hre.upgrades.upgradeProxy(upgradableProxyAddress, boxV2ContractFactory);
+await boxV2.waitForDeployment();
+```
+
+::
+
+### Custom deployment scripts with `hardhat-zksync`
+
+Additionally, you can write custom deployment scripts for ZKsync leveraging the `hre.deployer` object which is injected automatically by `@matterlabs/hardhat-zksync-deploy`.
+The `Deployer` class provides helper methods to deploy smart contracts to ZKsync. Find [more info here](https://docs.zksync.io/build/tooling/hardhat/hardhat-zksync-deploy#methods).
```
Comment (itsacoyote) [inline content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md:438]:
```suggestion
The `Deployer` class provides helper methods to deploy smart contracts to ZKsync. Learn more on [`hardhat-zksync-deploy` methods](/build/tooling/hardhat/hardhat-zksync-deploy#methods).
```

File: content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md [hunk 5]
```diff
@@ -1,275 +1,527 @@
...
+
+### Troubleshoting
+
+- **Contract size too large:** if the size of the generated bytecode is too large and can not be deployed, try compiling
+the contract with the `mode: 3` flag in the `hardhat.config.ts` file to
+optimize the bytecode size on compilation. [More info here](https://docs.zksync.io/build/tooling/hardhat/hardhat-zksync-solc#configuration).
```
Comment (itsacoyote) [inline content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md:444]:
```suggestion
optimize the bytecode size on compilation. Learn more on [`hardhat-zksync-solc` configuration](/build/tooling/hardhat/hardhat-zksync-solc#configuration).
```

File: content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md [hunk 6]
```diff
@@ -1,275 +1,527 @@
...
+
+## Smart Contract Verification
 
-Apart from the same classes and methods provided by ethers, zksync-ethers includes additional methods for zksync-specific features.
+<!-- TODO: pending rephrase -->
+For ZKsync, the verification of a smart contracts does not differ from the commonly used verification process.
```
Comment (itsacoyote) [inline content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md:449]:
```suggestion
There are no differences in the verification of smart contracts on ZKsync.
```

File: content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md [hunk 7]
```diff
@@ -1,275 +1,527 @@
...
+Find more info in [this section of the documentation](/build/tooling/hardhat/hardhat-zksync-verify).
+
+## Scripting
+
+<!-- TODO: Include mention to API compatibility in zksync-ethers and viem -->
+To be included.
+
+## Multichain projects
+
+Here are some recommendations projects that target multiple chains:
```
Comment (itsacoyote) [inline content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md:495]:
```suggestion
Here are some recommendations for projects that target multiple chains:
```

File: content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md [hunk 8]
```diff
@@ -1,275 +1,527 @@
...
+
+- Add the desired ZKsync networks with the `zksync:true` flag to the `hardhat.config.ts`.
+- Make sure to run the compilation task with the `--network` flag when targeting ZKsync networks to use the custom compiler.
+- The `@matterlabs/hardhat-zksync` plugin overrides the following plugins: `@nomiclabs/hardhat-ethers`, `@openzeppelin/hardhat-upgrades` when
+targeting ZKsync chains, while maintaining their original features when targeting EVM networks.
```
Comment (itsacoyote) [inline content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md:500]:
```suggestion
- When targeting ZKsync chains, the `@matterlabs/hardhat-zksync` plugin overrides the following plugins: `@nomiclabs/hardhat-ethers`, `@openzeppelin/hardhat-upgrades`.
```

File: content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md [hunk 9]
```diff
@@ -1,275 +1,527 @@
...
+- To avoid typescript collision errors between `@nomiclabs/hardhat-ethers` and
+`@openzeppelin/hardhat-upgrades` with  `@matterlabs/hardhat-zksync` make sure that only the latter is imported in the `hardhat.config.ts` file.
+- If you're using `hardhat-deploy` or `hardhat-ethers` your deployment scripts should work out of the box.
```
Comment (itsacoyote) [inline content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md:503]:
```suggestion
- Your deployment scripts should not require any changes if you're using `hardhat-deploy` or `hardhat-ethers`.
```

File: content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md [hunk 10]
```diff
@@ -1,275 +1,527 @@
...
+    },
+    zksyncTestnet: {
+      url: "https://sepolia.era.zksync.dev",
+      ethNetwork: "sepolia",
+      // ADDITION
+      deployPaths: "deploy-zksync", //single deployment directory
+      deployPaths: ["deploy", "deploy-zksync"], //multiple deployment directories
+      zksync: true,
+    },
+  },
```
Comment (itsacoyote) [inline content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md:521]:
```suggestion
  networks: {
    zksyncTestnet: {
      url: "https://sepolia.era.zksync.dev",
      ethNetwork: "sepolia",
      zksync: true,
      // ADDITION
      deployPaths: "deploy-zksync", //single deployment directory
      deployPaths: ["deploy", "deploy-zksync"], //multiple deployment directories
    },
  },
```

File: content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md [hunk 11]
```diff
@@ -1,275 +1,527 @@
...
+
+- Add the desired ZKsync networks with the `zksync:true` flag to the `hardhat.config.ts`.
+- Make sure to run the compilation task with the `--network` flag when targeting ZKsync networks to use the custom compiler.
+- The `@matterlabs/hardhat-zksync` plugin overrides the following plugins: `@nomiclabs/hardhat-ethers`, `@openzeppelin/hardhat-upgrades` when
+targeting ZKsync chains, while maintaining their original features when targeting EVM networks.
+- To avoid typescript collision errors between `@nomiclabs/hardhat-ethers` and
+`@openzeppelin/hardhat-upgrades` with  `@matterlabs/hardhat-zksync` make sure that only the latter is imported in the `hardhat.config.ts` file.
+- If you're using `hardhat-deploy` or `hardhat-ethers` your deployment scripts should work out of the box.
+- If you have different deployment scripts for
+ZKsync, you can indicate the custom deployment folder for the ZKsync network using [the `deployPaths` property](/build/tooling/hardhat/hardhat-zksync-deploy#network-specific-deployment-paths):
```
Comment (itsacoyote) [inline content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md:505]:
```suggestion
- If you have a separate directory for ZKsync deployment scripts, you can indicate the custom deployment folder for the ZKsync network using [the `deployPaths` property](/build/tooling/hardhat/hardhat-zksync-deploy#network-specific-deployment-paths):
```

**Review by itsacoyote — CHANGES_REQUESTED** [review]

File: content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md [hunk 12]
```diff
@@ -1,275 +1,527 @@
...
+        // Other networks
 
-## Overview
+    }
+    ```
 
-ZKsync Era offers [multiple Hardhat plugins](/build/tooling/hardhat/getting-started) with different features.
-This guide details the one you need to migrate your project to ZKsync Era.
+::callout{icon="i-heroicons-light-bulb"}
+You can also add the `zksync:true` flag to the `hardhat` network.
```
[the remark this answers is not in the document]
Comment (uF4No) [reply content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md:75]: It's optional. Only if you want to run tests with era-test-node

File: content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md (hunk 2 above)
Comment (uF4No) [reply content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md:300]: yes. I want to highlight that it's the same in EVM and ZKsync

File: content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md (hunk 3 above)
Comment (uF4No) [reply content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md:356]: with the new version of the plugins, yes

**Review by dutterbutter — APPROVED** [review]
LGTM

**Review by Romsters — APPROVED** [review]

**Review by itsacoyote — APPROVED** [review]

**Pull request merged**